// update-package-dependencies/lib/progress-element.js
(function() {
  var ProgressElement,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  ProgressElement = (function(_super) {
    __extends(ProgressElement, _super);

    function ProgressElement() {
      return ProgressElement.__super__.constructor.apply(this, arguments);
    }

    ProgressElement.prototype.createdCallback = function() {
      return this.tabIndex = -1;
    };

    ProgressElement.prototype.displayLoading = function() {
      return this.innerHTML = "<span class=\"loading loading-spinner-small inline-block\"></span>\n<span>\n  Updating package dependencies\u2026\n</span>";
    };

    ProgressElement.prototype.displaySuccess = function() {
      return this.innerHTML = "<span class=\"text-success\">\n  Package dependencies updated.\n</span>";
    };

    ProgressElement.prototype.displayFailure = function() {
      return this.innerHTML = "<span class=\"text-error\">\n  Failed to update package depencencies.\n</span>";
    };

    return ProgressElement;

  })(HTMLDivElement);

  module.exports = document.registerElement("update-package-dependencies-progress", {
    prototype: ProgressElement.prototype,
    "extends": "div"
  });

}).call(this);
