// service-hub/lib/provider.js
(function() {
  var CompositeDisposable, Provider, getValueAtKeyPath, semver, setValueAtKeyPath, _ref;

  CompositeDisposable = require('event-kit').CompositeDisposable;

  semver = require('semver');

  _ref = require('./helpers'), getValueAtKeyPath = _ref.getValueAtKeyPath, setValueAtKeyPath = _ref.setValueAtKeyPath;

  module.exports = Provider = (function() {
    function Provider(keyPath, version, service) {
      this.version = version;
      this.consumersDisposable = new CompositeDisposable;
      this.service = {};
      setValueAtKeyPath(this.service, keyPath, service);
    }

    Provider.prototype.provide = function(_arg) {
      var callback, consumerDisposable, keyPath, value, versionRange;
      keyPath = _arg.keyPath, versionRange = _arg.versionRange, callback = _arg.callback;
      if (semver.satisfies(this.version, versionRange)) {
        if (value = getValueAtKeyPath(this.service, keyPath)) {
          consumerDisposable = callback(value);
          if (typeof (consumerDisposable != null ? consumerDisposable.dispose : void 0) === 'function') {
            return this.consumersDisposable.add(consumerDisposable);
          }
        }
      }
    };

    Provider.prototype.destroy = function() {
      return this.consumersDisposable.dispose();
    };

    return Provider;

  })();

}).call(this);
