// status-bar/lib/launch-mode-view.js
(function() {
  var LaunchModeView,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  LaunchModeView = (function(_super) {
    __extends(LaunchModeView, _super);

    function LaunchModeView() {
      return LaunchModeView.__super__.constructor.apply(this, arguments);
    }

    LaunchModeView.prototype.initialize = function(_arg) {
      var devMode, safeMode, _ref;
      _ref = _arg != null ? _arg : {}, safeMode = _ref.safeMode, devMode = _ref.devMode;
      this.classList.add('inline-block', 'icon', 'icon-color-mode');
      if (devMode) {
        this.classList.add('text-error');
        return this.tooltipDisposable = atom.tooltips.add(this, {
          title: 'This window is in dev mode.'
        });
      } else if (safeMode) {
        this.classList.add('text-success');
        return this.tooltipDisposable = atom.tooltips.add(this, {
          title: 'This window is in safe mode.'
        });
      }
    };

    LaunchModeView.prototype.detachedCallback = function() {
      var _ref;
      return (_ref = this.tooltipDisposable) != null ? _ref.dispose() : void 0;
    };

    return LaunchModeView;

  })(HTMLElement);

  module.exports = document.registerElement('status-bar-launch-mode', {
    prototype: LaunchModeView.prototype,
    "extends": 'span'
  });

}).call(this);
