// atom-keymap/lib/helpers.js
(function() {
  var AtomModifierRegex, AtomModifiers, KeyboardEventModifiers, NumPadToASCII, SpecificityCache, WindowsAndLinuxCharCodeTranslations, WindowsAndLinuxKeyIdentifierTranslations, charCodeFromKeyIdentifier, fs, isASCII, keyFromCharCode, loophole, modifier, normalizeKeystroke, numpadToASCII, parseKeystroke, parser, pegjs, specificity, translateCharCodeForWindowsAndLinuxChromiumBug, translateKeyIdentifierForWindowsAndLinuxChromiumBug, _i, _j, _len, _len1, _ref, _ref1, _ref2;

  specificity = require('clear-cut').specificity;

  _ref = [], parser = _ref[0], fs = _ref[1], loophole = _ref[2], pegjs = _ref[3];

  AtomModifiers = new Set;

  _ref1 = ['ctrl', 'alt', 'shift', 'cmd'];
  for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
    modifier = _ref1[_i];
    AtomModifiers.add(modifier);
  }

  AtomModifierRegex = /(ctrl|alt|shift|cmd)$/;

  KeyboardEventModifiers = new Set;

  _ref2 = ['Control', 'Alt', 'Shift', 'Meta'];
  for (_j = 0, _len1 = _ref2.length; _j < _len1; _j++) {
    modifier = _ref2[_j];
    KeyboardEventModifiers.add(modifier);
  }

  SpecificityCache = {};

  WindowsAndLinuxKeyIdentifierTranslations = {
    'U+00A0': 'Shift',
    'U+00A1': 'Shift',
    'U+00A2': 'Control',
    'U+00A3': 'Control',
    'U+00A4': 'Alt',
    'U+00A5': 'Alt',
    'Win': 'Meta'
  };

  WindowsAndLinuxCharCodeTranslations = {
    48: {
      shifted: 41,
      unshifted: 48
    },
    49: {
      shifted: 33,
      unshifted: 49
    },
    50: {
      shifted: 64,
      unshifted: 50
    },
    51: {
      shifted: 35,
      unshifted: 51
    },
    52: {
      shifted: 36,
      unshifted: 52
    },
    53: {
      shifted: 37,
      unshifted: 53
    },
    54: {
      shifted: 94,
      unshifted: 54
    },
    55: {
      shifted: 38,
      unshifted: 55
    },
    56: {
      shifted: 42,
      unshifted: 56
    },
    57: {
      shifted: 40,
      unshifted: 57
    },
    186: {
      shifted: 58,
      unshifted: 59
    },
    187: {
      shifted: 43,
      unshifted: 61
    },
    188: {
      shifted: 60,
      unshifted: 44
    },
    189: {
      shifted: 95,
      unshifted: 45
    },
    190: {
      shifted: 62,
      unshifted: 46
    },
    191: {
      shifted: 63,
      unshifted: 47
    },
    192: {
      shifted: 126,
      unshifted: 96
    },
    219: {
      shifted: 123,
      unshifted: 91
    },
    220: {
      shifted: 124,
      unshifted: 92
    },
    221: {
      shifted: 125,
      unshifted: 93
    },
    222: {
      shifted: 34,
      unshifted: 39
    }
  };

  NumPadToASCII = {
    79: 47,
    74: 42,
    77: 45,
    75: 43,
    78: 46,
    96: 48,
    65: 49,
    66: 50,
    67: 51,
    68: 52,
    69: 53,
    70: 54,
    71: 55,
    72: 56,
    73: 57
  };

  exports.normalizeKeystrokes = function(keystrokes) {
    var keystroke, normalizedKeystroke, normalizedKeystrokes, _k, _len2, _ref3;
    normalizedKeystrokes = [];
    _ref3 = keystrokes.split(/\s+/);
    for (_k = 0, _len2 = _ref3.length; _k < _len2; _k++) {
      keystroke = _ref3[_k];
      if (normalizedKeystroke = normalizeKeystroke(keystroke)) {
        normalizedKeystrokes.push(normalizedKeystroke);
      } else {
        return false;
      }
    }
    return normalizedKeystrokes.join(' ');
  };

  exports.keystrokeForKeyboardEvent = function(event, dvorakQwertyWorkaroundEnabled) {
    var charCode, key, keyIdentifier, keystroke;
    keyIdentifier = event.keyIdentifier;
    if (process.platform === 'linux' || process.platform === 'win32') {
      keyIdentifier = translateKeyIdentifierForWindowsAndLinuxChromiumBug(keyIdentifier);
    }
    if (!KeyboardEventModifiers.has(keyIdentifier)) {
      charCode = charCodeFromKeyIdentifier(keyIdentifier);
      if (dvorakQwertyWorkaroundEnabled && typeof charCode === 'number') {
        if (event.keyCode === 46) {
          charCode = 127;
        } else {
          charCode = event.keyCode;
        }
      }
      if (charCode != null) {
        if (process.platform === 'linux' || process.platform === 'win32') {
          charCode = translateCharCodeForWindowsAndLinuxChromiumBug(charCode, event.shiftKey);
        }
        if (event.location === KeyboardEvent.DOM_KEY_LOCATION_NUMPAD) {
          charCode = numpadToASCII(charCode);
        }
        if (!isASCII(charCode) && isASCII(event.keyCode)) {
          charCode = event.which;
        }
        key = keyFromCharCode(charCode);
      } else {
        key = keyIdentifier.toLowerCase();
      }
    }
    keystroke = '';
    if (event.ctrlKey) {
      keystroke += 'ctrl';
    }
    if (event.altKey) {
      if (keystroke) {
        keystroke += '-';
      }
      keystroke += 'alt';
    }
    if (event.shiftKey) {
      if (!/^[^A-Za-z]$/.test(key)) {
        if (keystroke) {
          keystroke += '-';
        }
        keystroke += 'shift';
      }
      if (/^[a-z]$/.test(key)) {
        key = key.toUpperCase();
      }
    } else {
      if (/^[A-Z]$/.test(key)) {
        key = key.toLowerCase();
      }
    }
    if (event.metaKey) {
      if (keystroke) {
        keystroke += '-';
      }
      keystroke += 'cmd';
    }
    if (key != null) {
      if (keystroke) {
        keystroke += '-';
      }
      keystroke += key;
    }
    return keystroke;
  };

  exports.calculateSpecificity = function(selector) {
    return SpecificityCache[selector] != null ? SpecificityCache[selector] : SpecificityCache[selector] = specificity(selector);
  };

  exports.isAtomModifier = function(keystroke) {
    return AtomModifiers.has(keystroke) || AtomModifierRegex.test(keystroke);
  };

  exports.keydownEvent = function(key, _arg) {
    var alt, bubbles, cancelable, cmd, ctrl, event, keyCode, keyIdentifier, location, shift, target, view, _ref3;
    _ref3 = _arg != null ? _arg : {}, ctrl = _ref3.ctrl, shift = _ref3.shift, alt = _ref3.alt, cmd = _ref3.cmd, keyCode = _ref3.keyCode, target = _ref3.target, location = _ref3.location;
    event = document.createEvent('KeyboardEvent');
    bubbles = true;
    cancelable = true;
    view = null;
    if (/^[a-z]$/.test(key)) {
      key = key.toUpperCase();
    }
    if (key.length === 1) {
      keyIdentifier = "U+" + (key.charCodeAt(0).toString(16));
    } else {
      switch (key) {
        case 'ctrl':
          keyIdentifier = 'Control';
          ctrl = true;
          break;
        case 'alt':
          keyIdentifier = 'Alt';
          alt = true;
          break;
        case 'shift':
          keyIdentifier = 'Shift';
          shift = true;
          break;
        case 'cmd':
          keyIdentifier = 'Meta';
          cmd = true;
          break;
        default:
          keyIdentifier = key[0].toUpperCase() + key.slice(1);
      }
    }
    if (location == null) {
      location = KeyboardEvent.DOM_KEY_LOCATION_STANDARD;
    }
    event.initKeyboardEvent('keydown', bubbles, cancelable, view, keyIdentifier, location, ctrl, alt, shift, cmd);
    if (target != null) {
      Object.defineProperty(event, 'target', {
        get: function() {
          return target;
        }
      });
    }
    Object.defineProperty(event, 'keyCode', {
      get: function() {
        return keyCode;
      }
    });
    Object.defineProperty(event, 'which', {
      get: function() {
        return keyCode;
      }
    });
    return event;
  };

  normalizeKeystroke = function(keystroke) {
    var i, key, keys, modifiers, primaryKey, _k, _len2;
    keys = parseKeystroke(keystroke);
    primaryKey = null;
    modifiers = new Set;
    for (i = _k = 0, _len2 = keys.length; _k < _len2; i = ++_k) {
      key = keys[i];
      if (AtomModifiers.has(key)) {
        modifiers.add(key);
      } else {
        if (i === keys.length - 1) {
          primaryKey = key;
        } else {
          return false;
        }
      }
    }
    if (/^[A-Z]$/.test(primaryKey)) {
      modifiers.add('shift');
    }
    if (modifiers.has('shift') && /^[a-z]$/.test(primaryKey)) {
      primaryKey = primaryKey.toUpperCase();
    }
    keystroke = [];
    if (modifiers.has('ctrl')) {
      keystroke.push('ctrl');
    }
    if (modifiers.has('alt')) {
      keystroke.push('alt');
    }
    if (modifiers.has('shift')) {
      keystroke.push('shift');
    }
    if (modifiers.has('cmd')) {
      keystroke.push('cmd');
    }
    if (primaryKey != null) {
      keystroke.push(primaryKey);
    }
    return keystroke.join('-');
  };

  parseKeystroke = function(keystroke) {
    var e, keystrokeGrammar;
    if (parser == null) {
      try {
        parser = require('./keystroke');
      } catch (_error) {
        e = _error;
        if (fs == null) {
          fs = require('fs');
        }
        if (loophole == null) {
          loophole = require('loophole');
        }
        if (pegjs == null) {
          pegjs = require('pegjs');
        }
        keystrokeGrammar = fs.readFileSync(require.resolve('./keystroke.pegjs'), 'utf8');
        loophole.allowUnsafeEval((function(_this) {
          return function() {
            return parser = pegjs.buildParser(keystrokeGrammar);
          };
        })(this));
      }
    }
    return parser.parse(keystroke);
  };

  charCodeFromKeyIdentifier = function(keyIdentifier) {
    if (keyIdentifier.indexOf('U+') === 0) {
      return parseInt(keyIdentifier.slice(2), 16);
    }
  };

  translateKeyIdentifierForWindowsAndLinuxChromiumBug = function(keyIdentifier) {
    var _ref3;
    return (_ref3 = WindowsAndLinuxKeyIdentifierTranslations[keyIdentifier]) != null ? _ref3 : keyIdentifier;
  };

  translateCharCodeForWindowsAndLinuxChromiumBug = function(charCode, shift) {
    var translation;
    if (translation = WindowsAndLinuxCharCodeTranslations[charCode]) {
      if (shift) {
        return translation.shifted;
      } else {
        return translation.unshifted;
      }
    } else {
      return charCode;
    }
  };

  keyFromCharCode = function(charCode) {
    switch (charCode) {
      case 8:
        return 'backspace';
      case 9:
        return 'tab';
      case 13:
        return 'enter';
      case 27:
        return 'escape';
      case 32:
        return 'space';
      case 127:
        return 'delete';
      default:
        return String.fromCharCode(charCode);
    }
  };

  isASCII = function(charCode) {
    return (0 <= charCode && charCode <= 127);
  };

  numpadToASCII = function(charCode) {
    var _ref3;
    return (_ref3 = NumPadToASCII[charCode]) != null ? _ref3 : charCode;
  };

}).call(this);
